File: numbers/src/main/java/org/dicio/numbers/parser/lexer/Token.java
package org.dicio.numbers.parser.lexer;

import org.dicio.numbers.unit.Number;

public class Token {

    private static final Token EMPTY_TOKEN = new Token("", "");


    private final String value;
    private String spacesFollowing;

    Token(final String value, final String spacesFollowing) {
        this.value = value;
        this.spacesFollowing = spacesFollowing;
    }

    void setSpacesFollowing(final String spacesFollowing) {
        this.spacesFollowing = spacesFollowing;
    }


    public final String getValue() {
        return value;
    }

    public final String getSpacesFollowing() {
        return spacesFollowing;
    }

    public final boolean isValue(final String value) {
        return this.value.equalsIgnoreCase(value);
    }

    public Number getNumber() {
        throw new UnsupportedOperationException(); // throw by default, overridden
    }

    public boolean isNumberEqualTo(final long integer) {
        return false; // false by default, overridden
    }

    public boolean hasCategory(final String category) {
        return false; // false by default, overridden
    }

    public boolean isDurationToken() {
        return false; // false by default, overridden
    }

    public DurationToken asDurationToken() {
        throw new UnsupportedOperationException(); // throw by default, overridden
    }


    public static Token emptyToken() {
        return EMPTY_TOKEN;
    }
}

File: numbers/src/main/java/org/dicio/numbers/parser/lexer/Token.kt
package org.dicio.numbers.parser.lexer

import org.dicio.numbers.unit.Number

/**
 * @param positionInOriginalString index in the original string at which `value` starts
 */
open class Token internal constructor(
    @JvmField val value: String,
    @JvmField val spacesFollowing: String,
    @JvmField val positionInOriginalString: Int,
) {

    fun isValue(value: String?): Boolean {
        return this.value.equals(value, ignoreCase = true)
    }

    open fun isNumberEqualTo(integer: Long): Boolean {
        return false // false by default, overridden
    }

    open fun hasCategory(category: String): Boolean {
        return false // false by default, overridden
    }

    open val number: Number? = null

    open val asDurationToken: DurationToken? = null // null by default, but overridden


    companion object {
        private val EMPTY_TOKEN = Token("", "", -1)

        @JvmStatic
        fun emptyToken(): Token {
            return EMPTY_TOKEN
        }
    }
}
